### DISCUSSION TITLE
v4: Lua API & Template-based reports?

### DISCUSSION THREAD

**Discussion by rkusa** [rkusa/DATIS#113] [Ideas]
I am looking for feedback about the following changes to DATIS I am considering to make. Let me know if you'd welcome them, or if you think they'd just complicate things and you'd rather keep DATIS as it is right now.

---

tldr: I am thinking about replacing the current way of setting up DATIS stations with a Lua API and make the reports template-based for customization.

The current format:

```
ATIS Kutaisi 251.000, TRAFFIC 252.000, INFO Q, ACTIVE 21L, NO QFE, NO HPA
```

doesn't scale well with all the possible use-cases and customizations DATIS user's wish for.

So I am playing with the idea of completely removing the current approaches (text pattern inside the mission situation or inside of unit names) of setting up a station in the next major DATIS release.

I am thinking about providing a Lua API as a replacement. This would require adding the following code to `Scripts/MissionScripting.lua` to work:

```lua
-- DATIS
package.cpath = package.cpath..";"..lfs.writedir()..[[Mods\services\DATIS\bin\?.dll;]]
_G.DATIS = require('datis')
```

A mission editor would then setup stations like so (just some ideas, not the final API):

```lua
DATIS.new(251):airfield("Batumi"):start()

local custom = DATIS.new(128):template("custom message with wind heading {{wind_heading}} ..."):start()
custom:stop()
```

This would solve: #101, #99, #96 and #60 

I am also considering to make the report template-based. With that, everyone would be able to customize the report, and it is not really necessary to distinguish between ATIS, Weather station, Carrier and Broadcast stations anymore, because they'd then all simply use different templates.

The template (using handlebars syntax) for the current ATIS report would look like:

```xml
This is {{airfield_name}} information {{letter}}.
{{#if active_runway}}Runway in use is {{active_runway}}.
{{/if}}{{#if traffic_frequency}}Traffic frequency {{traffic_frequency}}.
{{/if}}{{#if spoken}}<phoneme alphabet="ipa" ph="w&#618;nd">Wind</phoneme>{{else}}Wind{{/if}} {{wind_heading}} at {{wind_speed_in_knots}} knots.
{{#if cloud_ceiling}}Ceiling {{cloud_ceiling}} {{cloud_coverage}}.
{{/if}}{{#if weather_conditions}}{{#each weather_conditions}}{{this}}{{#unless @last}},{{/unless}}{{/each}}.
{{/if}}{{#if visibility}}Visibility {{visibility}}.
{{/if}}Temperature {{temperature_in_celsius}} celsius.
ALTIMETER {{qnh_in_inhg}}.
REMARKS.
{{qnh_in_hpa}} hectopascal.
QFE {{qfe_in_inhg}} {{#if spoken}}<break time="500ms" /> {{/if}}or {{qfe_in_hpa}}.
End information {{letter}}.
```

DATIS would have several templates built-in: ATIS, Weather and Carrier.

This would solve: #104 (and similar requests I've received via Mail or Discord)

**Comment by poulet67** [comment]
I like it!

**Reply by adamburrito** [reply]
I'll chime in too... I think the more customization the better, so this looks fantastic! Would we be able to rearrange the order of everything in the ATIS report too? For example, I'd prefer to make it similar to a USA ATIS: have the visibility reported before clouds and runway in use (and/or approaches in use) after the weather information. And another thing to maybe consider, ATIS reports also include the dewpoint after the temperature (important information pertaining to moisture content in the atmosphere and possibility of fog either forming or perhaps predicting when it will be dissipating). Right now, DCS doesn't support dewpoint, but I'd bet they're planning on it in the future with the overhaul of the weather system. Maybe consider putting a place holder there now for it?

**Reply by rkusa** [reply]
@adamburrito Thanks for the feedback! Yes, you could re-arrange, trim, extend the message however you like. The only restrictions would be the available "variables" DATIS provides - eg: `{{cloud_ceiling}}`, `{{visibility}}`, etc.. But it wouldn't care if and where you'd put those "variables"/placeholders inside of your custom message. I think I'd rather wait until DCS actually exposes the drewpoint than putting in a static placeholder now. The proposed update would allow you to add a static drewpoint for your custom message though.

**Reply by adamburrito** [reply]
Awesome, very excited for this update, keep up the great work!

**Reply by HILOK** [reply]
thanks for all your good work, rkusa!
although i embrace the flexibility that a lua code approach would add, i would still vote to conserve the alternative simplistic way of adding a line to the situation briefing, because it can be done on the fly, and is transparent to anyone using the mission. if it means that the complete feature set is only available via lua, that's fine to me.
